src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs

```diff
@@ -356,73 +349,60 @@ internal static void RunFromThreadPoolDispatchLoop(ExecutionContext executionCon
             edi?.Throw();
         }
 
+
+        // For the two RunForThreadPoolUserWorkItem{WithDefault} methods: we aren't running in try/catch as if an exception 
+        // is directly thrown on the ThreadPool either process will crash or its a ThreadAbortException. 
+        // In either case sending AsyncLocal change notifications is unimportant.
         internal static void RunForThreadPoolUserWorkItem<TState>(ExecutionContext executionContext, Action<TState> callback, in TState state)
         {
-            Debug.Assert(Thread.CurrentThread.IsThreadPoolThread);
-            Debug.Assert(Thread.CurrentThread.ExecutionContext == null, "ThreadPool thread not on Default ExecutionContext.");
-            Debug.Assert(Thread.CurrentThread.SynchronizationContext == null, "ThreadPool thread not on Default SynchronizationContext.");
+            CheckThreadPoolAndContextsAreDefault();
+            Debug.Assert(executionContext != null && !executionContext.m_isDefault, "ExecutionContext argument is Default.");
 
-            // Capture current thread to local rather than looking it up multiple times
             Thread currentThread = Thread.CurrentThread;
-            if (executionContext != null && !executionContext.m_isDefault)
+            // Restore Non-Default context
+            currentThread.ExecutionContext = executionContext;
+            if (executionContext.HasChangeNotifications)
             {
-                // Restore changed ExecutionContext
-                currentThread.ExecutionContext = executionContext;
-                if (executionContext.HasChangeNotifications)
-                {
-                    // There are change notifications; trigger any affected
-                    OnValuesChanged(previousExecutionCtx: null, executionContext);
-                }
+                OnValuesChanged(previousExecutionCtx: null, executionContext);
             }
 
-            // Note: we aren't running in try/catch as if an exception is directly thrown on the threadpool
-            // either process will crash or its a ThreadAbortException. 
-            // In either case sending AsyncLocal change notifications is unimportant.
             callback.Invoke(state);
 
             ExecutionContext currentExecutionCtx = currentThread.ExecutionContext;
-            if (currentExecutionCtx != null)
+            if (currentExecutionCtx.HasChangeNotifications)
```

Comment:
Couldn't the callback have changed the execution context back to default (null)?

Reply:
No, as the only way to get back to the default context is being returned via the `ExecutionContext.Run{Variants}`; setting the last `AsyncLocal<T>` to `null` will leave you on a non-default context with an `EmptyAsyncLocalValueMap` as the locals